cpp/lib/include/MarcRecord.h

```diff
@@ -0,0 +1,166 @@
+/* 
+ * File:   MarcRecord.h
+ * Author: quboo01
+ *
+ * Created on 12. September 2016, 11:28
+ */
+
+#ifndef MARCRECORD_H
+#define MARCRECORD_H
+
+#include "Compiler.h"
+#include "DirectoryEntry.h"
+#include "File.h"
+#include "Leader.h"
+#include "Subfields.h"
+#include "XmlWriter.h"
+
+#include <string>
+#include <iostream>
+#include <set>
+#include <unordered_set>
+#include <vector>
+
+
+class MarcRecord {
+    friend class MarcReader;
+    friend class MarcWriter;
+public:
+    static const size_t FIELD_NOT_FOUND = (size_t) -1;
+private:
+    Leader leader_;
+    std::vector<DirectoryEntry> directory_entries_;
+    std::string raw_data_;
+    std::set<size_t> deleted_field_indices;
+
+    MarcRecord() = default;
+    MarcRecord(Leader &leader, std::vector<DirectoryEntry> directory_entries, std::string &raw_data)
+        : leader_(std::move(leader)), directory_entries_(std::move(directory_entries)), raw_data_(std::move(raw_data)) {}
+
+public:
+    MarcRecord(MarcRecord &&other) noexcept
+    : leader_(std::move(other.leader_)), directory_entries_(std::move(other.directory_entries_)),
+      raw_data_(std::move(other.raw_data_)) { }
+
+    operator bool () const { return not directory_entries_.empty(); }
+    const Leader &getLeader() const { return leader_; }
+    Leader &getLeader() { return leader_; }
+
+    Leader::RecordType getRecordType() const { return leader_.getRecordType(); }
+    std::string getControlNumber() const { return getFieldData("001"); }
+
+    size_t getNumberOfFields() const { return directory_entries_.size(); }
+    std::string getFieldData(const std::string &tag) const;
+    std::string getFieldData(const size_t tag_index) const;
+    Subfields getSubfields(const std::string &tag) const;
+    Subfields getSubfields(const size_t field_index) const;
+
+    std::string getTag(const size_t index) const;
+
+    /** \brief Returns the index of "field_tag" or -1 if the tag is not present. */
+    size_t getFieldIndex(const std::string &field_tag) const;
+
+    /** \return The number of field indices for the tag "tag". */
+    size_t getFieldIndices(const std::string &field_tag, std::vector<size_t> * const field_indices) const;
+    
+    /** \brief Updates the field at index "field_index" and adjusts various field and records lengths. */
+    bool updateField(const size_t field_index, const std::string &new_field_contents);
+
+    size_t insertField(const std::string &new_field_tag, const std::string &new_field_value);
+
+    /** \brief Deletes the field at index "field_index" and adjusts various field and records lengths. */
+    void deleteField(const size_t field_index);
+
+    void markFieldAsDeleted(const size_t field_index);
```

Comment:
This seems to duplicate the functionality of the previous function.  I suspect we can do w/o it.